lambda/transl_list_comprehension.ml

```diff
@@ -0,0 +1,302 @@
+      in
+      let inner_acc = Ident.create_local "accumulator" in
+      let body_arg_lets, body =
+        translate_bindings
+          ~transl_exp ~scopes ~loc
+          ~inner_body ~accumulator:(Lvar inner_acc) bindings
+      in
+      let body_func =
+        (* CR aspectorzabusky: How do I make this function stack allocate its result? *)
+        Lfunction
+          { kind   = Curried { nlocal = 1 } (* The accumulator is local *)
+          ; params = [element, element_kind; inner_acc, Pgenval]
+          ; return = Pgenval
+          ; attr   = default_function_attribute
+          ; loc
+          ; mode   = alloc_local
+          ; region = true
+              (* CR aspectorzabusky: We want one region per iterator, right? *)
+          ; body   = add_bindings body
+          }
+      in
+      let result =
+        Lambda_utils.apply
+          ~loc
+          (Lazy.force builder)
+          (List.map (fun Let_binding.{id; _} -> Lvar id) arg_lets @
+           [body_func; accumulator])
+      in
+      arg_lets @ body_arg_lets, result
+  | [] ->
+      [], inner_body ~accumulator
+
+(** Translate a single clause, either [for ... and ...] or [when ...]
+    ([Typedtree.comprehension_clause]), into a single Lambda term of type
+    ['res rev_dlist], assuming we know how to translate everything that ought to
+    be nested within it (the [comprehension_body], a a function awaiting the
+    most nested accumulator as a labeled argument which will produce the body of
+    the iterations) and have a name for the accumulator of the current
+    [rev_dlist] ([accumulator], which changes at every recursive step). *)
+let rec translate_clauses
+          ~transl_exp ~scopes ~loc ~comprehension_body ~accumulator = function
+  | clause :: clauses ->
+      let body ~accumulator =
+        translate_clauses ~transl_exp ~scopes ~loc
+          ~comprehension_body ~accumulator clauses
+      in begin
+        match clause with
+        | Texp_comp_for bindings ->
+            let arg_lets, bindings =
+              translate_bindings
+                ~transl_exp ~scopes ~loc ~inner_body:body ~accumulator bindings
+            in
+            Let_binding.let_all arg_lets bindings
+        | Texp_comp_when cond ->
+            Lifthenelse(transl_exp ~scopes cond,
+                        body ~accumulator,
+                        accumulator,
+                        Pgenval (* [list]s have the standard representation *))
+      end
+  | [] ->
+      comprehension_body ~accumulator
+
+let comprehension ~transl_exp ~scopes ~loc { comp_body; comp_clauses } =
+  let rev_comprehension =
+    translate_clauses ~transl_exp ~scopes ~loc
+      ~comprehension_body:(fun ~accumulator ->
+        rev_list_snoc_local
+          ~loc
+          ~init:accumulator
+          ~last:(transl_exp ~scopes comp_body))
+      ~accumulator:rev_list_nil
+      comp_clauses
+  in
+  Lambda_utils.apply ~loc (Lazy.force rev_list_to_list) [rev_comprehension]
```

Comment:
Summary comment: This implementation feels rather less refined than the one for arrays, which struck me as very pleasant reading. For example, is continuation-passing style necessary here? What does it accomplish? It just somehow feels less polished to me.

Reply:
It is less polished; as you can tell, I wrote it first :-)  I think we agreed that it's not worth trying to totally refactor, though.